src/ECS/ECS.cpp

```diff
@@ -0,0 +1,179 @@
+        }
+        
+        void renderUnit (PositionComponent position, AnimatedComponent animationComponent, std::shared_ptr<Gm1File> assetFile) {
+            Camera& cam = Camera::instance();
+            auto [x, y] = position;
+
+            int px = x * 30 - cam.positionX;
+            int py = y * 15 - cam.positionY;
+
+            SDL_Rect r = assetFile->GetTextureAtlas()->Get(animationComponent.animate ? animationComponent.frameIndex : 0); // TODO: index
+            Rendering::Render(*assetFile->GetTextureAtlas(), px, py, &r);
+        }
+
+        void render (entt::registry &registry) {
+            auto view = registry.view<const EntityTypeComponent, const PositionComponent, const AnimatedComponent>();
+
+            // this is just getting pointers to preloaded, parsed file data
+            // but maybe it can be moved somewhere else if this is performance hit
+            auto lordFile = AssetStorage::getFileGM1(AssetStorageKey::BODY_LORD_GM1);
+            auto deerFile = AssetStorage::getFileGM1(AssetStorageKey::BODY_DEER_GM1);
+
+            auto appleTreeFile = AssetStorage::getFileGM1(AssetStorageKey::TREE_APPLE_GM1);
+            auto birchTreeFile = AssetStorage::getFileGM1(AssetStorageKey::TREE_BIRCH_GM1);
+            auto chestnutTreeFile = AssetStorage::getFileGM1(AssetStorageKey::TREE_CHESTNUT_GM1);
+            auto oakTreeFile = AssetStorage::getFileGM1(AssetStorageKey::TREE_OAK_GM1);
+            auto pineTreeFile = AssetStorage::getFileGM1(AssetStorageKey::TREE_PINE_GM1);
+            auto shrub1TreeFile = AssetStorage::getFileGM1(AssetStorageKey::TREE_SHRUB1_GM1);
+            auto shrub2TreeFile = AssetStorage::getFileGM1(AssetStorageKey::TREE_SHRUB2_GM1);
+
+            for(auto [entity, typeComponent, positionComponent, animationComponent]: view.each()) {
+                switch (typeComponent.type) {
+                    case EntityType::LORD: renderUnit(positionComponent, animationComponent, lordFile); break;
+                    case EntityType::DEER: renderUnit(positionComponent, animationComponent, deerFile); break;
+
+                    case EntityType::TREE_APPLE: renderUnit(positionComponent, animationComponent, appleTreeFile); break;
+                    case EntityType::TREE_BIRCH: renderUnit(positionComponent, animationComponent, birchTreeFile); break;
+                    case EntityType::TREE_CHESTNUT: renderUnit(positionComponent, animationComponent, chestnutTreeFile); break;
+                    case EntityType::TREE_OAK: renderUnit(positionComponent, animationComponent, oakTreeFile); break;
+                    case EntityType::TREE_PINE: renderUnit(positionComponent, animationComponent, pineTreeFile); break;
+                    case EntityType::TREE_SHRUB1: renderUnit(positionComponent, animationComponent, shrub1TreeFile); break;
+                    case EntityType::TREE_SHRUB2: renderUnit(positionComponent, animationComponent, shrub2TreeFile); break;
+                }
+            }
```

Comment:
This is very type intensive.
Figure out which parts need to be switched over and extract it into a seperate function or module.
For example:
```C++
auto getEntityAssetFile = [&](EntityType type){
  switch(type){return ....}
}
for(loop over entities ...){
 renderUnit(... getEntityAssetFile(type));
}
 
```

Reply:
Wow :D I've seen that before, but I had no idea how to practically use that :D

Reply:
I just realized, you can just put all of that in a map and access the gm1 with the type as a key.
I think that's easier, than using a switch.

Reply:
Sounds promising, will check that out after cleaning

Reply:
Sounds promising, will check that out after cleaning

Reply:
I think we've come to the conclusion there